### PR TITLE
Keyed VDom

### PR DESCRIPTION
**Pull request by zaceno**
Allow instructing the patch-function what dom nodes to reuse in update, via 

```
h(tag, {_key: '...', ...}, [])
```

### CODE REVIEW

**Comment by zaceno** [comment]
Solves https://github.com/hyperapp/hyperapp/issues/117 (among many other important use-cases)

Here is the code pen from that issue, demonstrating how it solves the problem. 
(wierdly though, something makes it break in safari, although it works in chrome. Need to look more at that before mergeing)

**Comment by zaceno** [comment]
Here is a nother, simpler problem this PR solves.

http://codepen.io/zaceno/pen/ryWddK

To demonstrate the problem, click the update button and notice how the images reload.
Now uncomment line 17 (_key: 'somekey') and try again. Notice that because the photo-gallery is rendered with the same _key property, the patch function knows to reuse it's dom-node, rather than replace it with the "updating..." message (and then create a new image-gallery node from scratch)

**Comment by leeoniya** [comment]
keyed nodes solve many problems, broadly summarized here:

https://github.com/hyperapp/hyperapp/issues/132#issuecomment-284704949

**Comment by farism** [comment]
It is definitely something that will be needed in some form or another for long term performance reasons.

I haven't looked at the implementation, but the first thing that jumps out at me is why the leading `_` in `_key`?

**Comment by leeoniya** [comment]
keys rarely help performance if implemented inefficiently can reduce performance. the point of keys is to prevent unintended node reuse and busting any intermittent dom state that's not explicitly managed by the vdom lib.

the `_` prefix is to discern vars used by the vdom lib itself from real dom attributes applied to the elements.

**Comment by farism** [comment]
Great, thank you for the explanation @leeoniya

**Comment by jbucaran** [comment]
@zaceno In your example what is `model.current`? Is it a number, a string? There is no model schema to infer this so one needs to understand all the code to figure it out.

**Comment by jbucaran** [comment]
@leeoniya Could you review this PR? 🙏

**Comment by jbucaran** [comment]
I'm going to need everyone's help to review this PR. @farism @dodekeract @danigb @selfup @jamen @liadbiz 🙏 🙏

**Comment by leeoniya** [comment]
i can sanity check it tomorrow. keep in mind though, an optimal keyed algo is thoroughly described here:

https://github.com/localvoid/kivi/blob/master/lib/vnode.ts#L1284

**Comment by zaceno** [comment]
@jbucaran: in my example model.current and model.next are strings with URLs

**Comment by zaceno** [comment]
Amen to some review on this! At this stage it's mostly a proof of concept that it's doable. But certainly needs some community-scrutiny

File: src/app.js [hunk 1]
```diff
@@ -275,12 +276,50 @@ export default function (app) {
...
+      var len = node.children.length, oldLen = oldChildren.length
+      for (var i = 0, j = 0; i < len || j < oldLen; i++) {
         var child = node.children[i]
+        //in case the next new node is keyed (and a previous node is known
+        //we want the element inserted at this index, and then patched
+        //normally, with the corresponding oldNode (which may be moved)
+
+        if (child && child.data && child.data._key && keyedChildren[child.data._key]) {
+          var keyed = keyedChildren[child.data._key]
+          element.appendChild(keyed.element)
```
Comment (zaceno) [inline src/app.js]: Probably more effincient in most cases, if this line were: `element.insertBefore(keyed.element, element.childNodes[i])`

**Comment by zaceno** [comment]
There is an issue with https://github.com/hyperapp/hyperapp/issues/141 and css transitions. It seems that the  way we remove and reinsert keyed elements in the dom somehow breaks css-transitions in Chrome and Safari. 

In order to make my sliding images demo work, I have to add classes to the element *after* the patch function is run, i e in oncreate and onupdate lifecycle methods.

http://codepen.io/zaceno/pen/dvOmNr

... not sure if there is something we should do about that in the patch-function, or just call it a browser-bug and accept the workaround.

**Comment by leeoniya** [comment]
quick update, because i've been somewhat busy. i want to be clear about what i plan check for this PR, and if anyone gets to this first, go ahead and do it.

for [obvious reasons](https://github.com/leeoniya/domvm), i don't have much interest in dedicating much time to hyperapp. i i spot anything obvious that i've learned to avoid, i will certainly point it out, but i will not be digging into the source here to validate that all things are coded optimally and without errors - that job is painstakingly huge, and i've already gone through it all in domvm.

that being said, here's an easy way to figure out whether this PR does what it's supposed to from external testing alone:

1. create a list of 5 sibling nodes via hyperapp, each should have a `textContent` of a different color like "red". each should also be `_key`ed by the same value.
2. use vanilla dom to add a `.style.color = "red"` to match the created nodes' textContent.
3. return various templates in hyperapp that represent the following mutations:
  a. remove first node
  b. remove last node
  c. remove a node from middle
  d. swap 2 adjacent middle nodes
  e. swap first and last nodes
  f. insert a node in front
  g. insert a node at end
  h. insert a node in middle
  i. insert 2 disjoint nodes in middle.
  j. remove 2 disjoint nodes from middle.
4. after each "redraw", use iterate the DOM and ensure that the textContent matches `.style.color`.
5. validate which DOM operations were performed to achieve the needed dom state. you can use mutationObservers for this or you can use a lib i made for testing this in domvm [1]. Youc an see how it's used in domvm's regression tests [2].

I don't have a timeline for when i can get around to this, life is quite hectic these days, sorry.

[1] https://github.com/leeoniya/domvm/blob/2.x-dev/test/dominstr.js
[2] https://github.com/leeoniya/domvm/blob/2.x-dev/test/index.html

**Comment by zaceno** [comment]
@leeoniya thanks for that excellent suite of tests. I couldn't help myself but started checking my implementation against the suite you recommended (even though I know self-verification is not real verification) -- and *it did* uncover bugs in the implementation that I will start looking at now.

So there's no rush for anyone to verify this just yet (unless you're curious what the bugs were, and interested in fixing them, of course ;) )

**Comment by leeoniya** [comment]
> (even though I know self-verification is not real verification)

that would imply that developers cannot be trusted to write unit tests for their own code 😆

**Comment by zaceno** [comment]
Ok, so the problem was the `element.appendChild...` line I commented on yesterday. It wasn't inefficcient -- it was *wrong* ;) After fixing that, @leeoniya 's tests all pass.

To "prove" it, here's a code-pen that cycles through the tests. One each second so that you can follow them visually.

If at any point the element.style.color doesn't match the element.textContent, the console should shout at you. Fortunately it doesn't :)

... Again, I prove nothing by verifying my own code of course. But consider it a green light to proceed with more verification & reviewing.

**Comment by zaceno** [comment]
@leeoniya ... hehe, you make a good point. But then I at least hope someone looks at my tests ;)

**Comment by leeoniya** [comment]
> Again, I prove nothing by verifying my own code of course. But consider it a green light to proceed with more verification & reviewing.

no implementation is bug free. but code that passes a non-trivial set of smell tests *and* has automated regression tests can be merged. if all existing tests pass and no new tests fail, that's the best anyone can do. the only enemy is not enough existing tests - there are never enough.

**Comment by zaceno** [comment]
I can't explain how a test failed in circleci from my most recent push. Locally they're all green!

Besides all I did was remove a commented-out line I'd left in by mistake. (But you wouldn't know, because I did a `git commit --amend` and then a `git push --force`)

**Comment by ngryman** [comment]
IMHO `_key` should be `key` (https://github.com/Matt-Esch/virtual-dom/tree/master/virtual-hyperscript#special-properties-in-h).

**Comment by zaceno** [comment]
@ngryman -- fine with me. Makes sense considering onremove et c also have no special prefix.

**Comment by jbucaran** [comment]
→ onRemove :)

**Comment by AutoSponge** [comment]
So, I put in https://github.com/hyperapp/hyperapp/pull/151. Based on my work with morphdom solving a very similar issue, my code should solve this if the "key" element implements `isSameNode` as an expando property.

Here's the code that changed:
```js
function shouldUpdate(a, b) {
    return a.tag !== b.tag || typeof a !== typeof b || isPrimitive(a) && a !== b || isElement(a) && !a.isSameNode(b)
}
```

By declaring a virtual node as the same as another (existing DOM node) with `isSameNode` you can make the renderer implementation "skip" adding/mutating the in-place node.

**Comment by jbucaran** [comment]
@AutoSponge Do you mean #151 is an alternative to adding key (this PR) or something else? I'm afraid I don't understand what you are saying.

**Comment by AutoSponge** [comment]
@jbucaran it made more sense in my head...

To allow for keys to work and allowing real elements to be held by reference in the model (my PR), the `setElementData` process would need to set `isSameNode` to the element (since this implementation only knows how to use keys based on `node.data.key` which would not exist on an element).  The `isSameNode` fn would return true when passed a vnode having the expected key.

**Comment by jbucaran** [comment]
@AutoSponge Haha okay.

I am currently working on my own keys implementation based in #141. Like you say I am not sure if it will be compatible with #151, but on the other hand, I still haven't quite understood what's the purpose of that PR. Don't take that as criticism please, on the contrary, I just need more time or perhaps if you have time you can try ELI5 to me again.

**Comment by AutoSponge** [comment]
@jbucaran no problem! I'll add explanation to #151

**Comment by jbucaran** [comment]
@zaceno 👋 Is this done or you planning tweaking something more?

**Comment by zaceno** [comment]
@jbucaran, Hi sorry for the silence :) 

No, it's not complete. I am working on a rewrite where I do removeChild(...) only when absolutely necessary (to fix the css-transition issue, among other things). It's going to make the diff bigger, but it will be a more complete solution.  

I didn't get as much time as I had hoped this weekend, but I am working away at it little by little, and hope to have an update to this PR soon. By the end of the week at the latest.

**Comment by zaceno** [comment]
@jbucaran So *now* I believe this solution is complete. It ended up bigger than I hoped, and perhaps touching more of the code than you would like. But it *works*.

(And because I don't needlessly remove & reinsert dom-nodes any longer, it does not break css transitions.)

Here are updated versions of all the code pens I've mentioned before:

@leeoniya s suite of smoke tests visualized here: http://codepen.io/zaceno/pen/wJqmjY
...notice that the console is silent. Proof that the keyed vdom solution does what it should

My css transitions slideshow with images: http://codepen.io/zaceno/pen/WpEJQV
... no flickering reloading images. See it break when you comment out the `key: ...` lines.

My css transitions slideshow with iframes: http://codepen.io/zaceno/pen/oZeqRG
... no reloading iframes! See it break when you comment out the `key: ...`lines.

the image gallery example: http://codepen.io/zaceno/pen/bqrvQO
... notice how when you press the update button, the gallery images *don't* reload (as intended). but if you comment out the `key: ...` line, then they *do* reload when the update button is pressed (a problem keys solve)

**Comment by AutoSponge** [comment]
I solved a similar problem in my PR #151. The problem comes when replacing an element with a sibling element instead of just removing the preceding sibling. 

I simply perform the remove then, remove the reference from the `oldNode.children` array then update the array's length, `oldLen`. If `oldNode.children` held a reference to the element it created the first time it went through render, you could perform the same check I'm doing (since I'm looking at real dom nodes and using `.compareDocumentPosition`. 

If you want, I can try to add that to my PR and test your _original_ demos against it.

**Comment by zaceno** [comment]
@AutoSponge, if I get what you're saying, I think a keyed vdom covers more cases than just the one you're describing, but the case you're describing is the problem I initially had with my iframe/image slidedshow demos.

I know my case can be solved with something simpler than keyed vdoms (I had a PR out before this one, that solved it in a more minimal way) -- but I'm absolutely curious to see your solution (I think I'd understand better what you're saying if I saw it anyway), so go for it! :)

That said, because keyed vdoms cover more cases (specifically: when you need to reuse but *reorder* dom elements), I'm hoping for that feature in hyperapp rather than a less complete solution.

Hope I understood you correctly! If I didn't, it wasnt intentional :)

**Comment by AutoSponge** [comment]
@zaceno I started this work but I'm not quite done. As I worked, I thought `id` should be used as the key instead of `key` since (a) it should already be unique and (b) it's easy to get the list of "keyed" elements with `parent.querySelectorAll("[id]")`. 

I could continue to work on this implementation but don't let me hold up acceptance. My changes in #151 are likely to fold in easily since it's not that big.

**Comment by zaceno** [comment]
@AutoSponge Cool, thanks for sharing that idea! The thing about keys vs ids is, though, is that while id's must be unique within the entire document, keys only need to be unique among sibling-nodes. And there are cases where you might want to have the same key in multiple places in a document. Like if you have multiple of the same components on a page.

By all means keep working, and don't hold your breath on this PR! I hope its not long before we can have keyed patching in hyperapp, but this may not be the implementation that makes it.

@jbucaran expressed hesitance on the slack-channel, against both the performance-degradation and increased bundle size this brings with it. So to make this happen I'll need to both optimize and condense it. But frankly, right now I'm not sure I can do any better, so I'm not feeling that hopeful :(

Thankfully @jbucaran is working on an implementation of his own, and I hope he fares better :)

**Comment by zaceno** [comment]
Given the state of this PR, I want to express my purely personal and subjective opinion that: if no better solution (more performant, smaller) can be found in the near term, this PR should be merged anyway. 

There are several use cases where this functionality (not necessarily my implementation) is absolutely critical. Hyperapp *needs* this if it wants to be a "general purpose" frontend framework. 

Despite the shortcomings of my implementation in terms of size and performance, it *does work*. Once it's in, we can and should continue looking for optimizations. 

... that is of course *only if* no better implementation can be found in the short term. I'm still willing to wait for @jbucaran 's solution (or anyone else currently having a go at it)

**Comment by jbucaran** [comment]
@zaceno Thanks for your work on this PR! I agree with you this is an important feature we need to ship ASAP. I hope I can dedicate some good time to this on the weekend.

**Comment by zaceno** [comment]
So I made some adjustments and brought back the batch-delete. (thanks @jbucaran for brilliantly reminding me that dom nodes don't have to be detached from the document to be inserted!!) I also tried to refactor it into a hopefully more compressible way.

 I wasn't able to get the benchmark thing running, so I havent been able to check performance. But hopefully this should be much better in performance, and a little smaller as well.

File: src/app.js [hunk 2]
```diff
@@ -215,48 +217,134 @@ export default function (app) {
...
+          element: element.childNodes[oldIndex + indexOffset]
+        }
+        oldIndex++
       }
     }
 
-    return element
+    for (var key in cache) {
+      var unused = cache[key]
+      batchRemove(element, unused.element, unused.node)
```
Comment (zaceno) [inline src/app.js]: I'm pretty sure that can't happen. Unkeyed nodes get scheduled for removal as we iterate over them. Keyed nodes are put in the cache, and scheduled for removal after we see we haven't used them.

File: test/app.test.js [hunk 3]
```diff
@@ -288,6 +288,209 @@ describe("app", () => {
...
+                } else {
+                  return h("ul", {onUpdate: checkColors}, test.children)
+                }
+              }
+            })
+          })
+        })
+      })
+
+      it("fires onremove for ONLY the not-reused keyed nodes", done => {
```
Comment (zaceno) [inline test/app.test.js]: Ok, I'll remove them. But can we wait until it's definitely time to merge? Because they're pretty useful for knowing I don't break stuff when I fix merge-conflicts.

Comment (jbucaran) [inline test/app.test.js]: Of course, of course.

File: src/app.js [hunk 4]
```diff
@@ -215,48 +217,134 @@ export default function (app) {
...
+    ) {
+      newChild = newNode.children[newIndex]
+      oldChild = oldNode.children[oldIndex]
+
+      if (
+        newChild !== undefined &&
+        isKeyed(newChild) &&
+        cache[newChild.data.key]
+      ) {
+        var c = cache[newChild.data.key]
```
Comment (zaceno) [inline src/app.js]: You mean `var i` compresses better than `var c`? Is that because `var i` is already used in a lot if places? 

Either way, I'll fix that, no problem.

File: src/app.js [hunk 5]
```diff
@@ -215,48 +217,134 @@ export default function (app) {
...
       }
     }
   }
 
-  function patch(parent, element, oldNode, node) {
-    if (oldNode === undefined) {
-      element = parent.appendChild(createElementFrom(node))
-
-    } else if (node === undefined) {
+  function batchRemove (parent, element, node) {
```
Comment (zaceno) [inline src/app.js]: I don't se any problem batching the onRemove-handler, BUT even before this PR, the onRemove handler was deferred, meaning it gets was called *after* the node is removed. So... do you want me to fix that behavior as well, or leave it as it is?

File: src/app.js [hunk 6]
```diff
@@ -215,48 +217,134 @@ export default function (app) {
...
-
-      if (oldNode && oldNode.data && oldNode.data.onRemove) {
-        defer(oldNode.data.onRemove, element)
+      if (node.data && node.data.onRemove) {
+        defer(node.data.onRemove, element)
       }
+  }
+
+  function patch (parent, element, oldNode, newNode) {
+    if (oldNode === undefined) {
```
Comment (zaceno) [inline src/app.js]: Yes, that is correct

File: src/app.js [hunk 7]
```diff
@@ -215,48 +217,134 @@ export default function (app) {
...
+      var n = createElementFrom(newNode)
+      parent.insertBefore(n, element)
+      element = n
 
     } else if (
```
Comment (zaceno) [inline src/app.js]: If you're referring to the "insert app directly into document.body" merge, then yes. This code has been updated specifically to adapt to that (to return the new element of a patch)

File: src/app.js [hunk 8]
```diff
@@ -215,48 +217,134 @@ export default function (app) {
...
+    }
+
+    return element
+  }
+
+  function patchChildren(element, oldNode, newNode) {
+    // We need to index all keyed nodes in oldChildren in order to
+    // distinguish new keyed nodes from nodes to reuse in newChildren
+    var keyedChildren = {}
+    for (var i = 0; i < oldNode.children.length; i++) {
```
Comment (zaceno) [inline src/app.js]: I'll try it :)

File: src/app.js [hunk 9]
```diff
@@ -215,48 +217,134 @@ export default function (app) {
...
+    var newChild
+
+    var patchChild = function () {
+      patch(element, element.childNodes[oldIndex + indexOffset], oldChild, newChild)
+      oldIndex++
+      newIndex++
+    }
+
+    var insertChild = function () {
+      patch(element, element.childNodes[oldIndex + indexOffset], undefined, newChild)
```
Comment (zaceno) [inline src/app.js]: Ok. But it means the variable declaration for `node` at the top of the file will have to be instantiated to null as well. Trying that.

Comment (zaceno) [inline src/app.js]: Actually no. can't use null here. The patcher must be prepared to accept undefined values for oldChild parameter. And we can't just check `if(!oldChild)` because it could be an empty string. So we'd have to check `if (oldChild === null || oldChild === undefined)` ... which defeats the idea of *removing* bytes :)

File: src/app.js [hunk 10]
```diff
@@ -215,48 +217,134 @@ export default function (app) {
...
+      newIndex++
+    }
+
+    var removeChild = function () {
+      batchRemove(element, element.childNodes[oldIndex + indexOffset], oldChild)
+      oldIndex++
+    }
+
+    while (
+      newNode.children[newIndex] !== undefined ||
```
Comment (zaceno) [inline src/app.js]: Totally. Don't know why I didn't think of that. Probably left over from some earlier attempt where the child-lists were mutated (so you couldn't rely on lengths). I'll fix this!

File: src/app.js (hunk 9 above)
Comment (zaceno) [inline src/app.js]: Actually *YES*... I just realized that `'' == null` is false and `undefined==null` is true. So we can use null here with no problems. :)

File: src/app.js [hunk 11]
```diff
@@ -215,48 +217,134 @@ export default function (app) {
...
+        removeChild()
+
+      } else if (!isKeyed(newChild)) {
+        insertChild()
+
+      } else if (oldChild.data.key === newChild.data.key) {
+        patchChild()
+
+      } else {
+        cache[oldChild.data.key] = {
```
Comment (zaceno) [inline src/app.js]: Hmm... not sure I follow. I'm happy to try it, but could you give me a code-snippet to explain more. What I'm envisioning is not smaller so I'm sure I'm misunderstanding :)

File: src/app.js (hunk 5 above)
Comment (jbucaran) [inline src/app.js]: I pushed it to master a while ago, it's like 10 bytes smaller so...

Comment (jbucaran) [inline src/app.js]: 😆

File: src/app.js [hunk 12]
```diff
@@ -215,48 +217,134 @@ export default function (app) {
...
+    var keyedChildren = {}
+    for (var i = 0; i < oldNode.children.length; i++) {
+      if (!oldNode.children[i].data) continue
+      if (!oldNode.children[i].data.key) continue
+      keyedChildren[oldNode.children[i].data.key] = true
+    }
 
-      var len = node.children.length
-      var oldLen = oldNode.children.length
+    var isKeyed = function (node) {
```
Comment (zaceno) [inline src/app.js]: Mostly for readability. Because they were repeated code snippets. But now that I reduced my 8-branch if-else down to its minimum 5-branches, I have got rid of those extra functions, and just inlined the code in the if-else.

Comment (zaceno) [inline src/app.js]: the `!!` was a dumb mistake. Fixed

I'll look more into how to avoid isKeyed!

**Comment by zaceno** [comment]
Latest adjustments brings the total compressed bundle size down to 1954 bytes.

File: src/app.js [hunk 13]
```diff
@@ -254,93 +254,66 @@ export default function (app) {
   function patchChildren(element, oldNode, newNode) {
     // We need to index all keyed nodes in oldChildren in order to
     // distinguish new keyed nodes from nodes to reuse in newChildren
-    var keyedChildren = {}
+    var oldChildren = []
```
Comment (jbucaran) [inline src/app.js]: @zaceno Nice!

File: src/app.js (hunk 12 above)
Comment (jbucaran) [inline src/app.js]: I think there should be no `var x = function(){..}`. This pattern only adds more bytes.

File: src/app.js [hunk 14]
```diff
@@ -211,40 +213,107 @@ export default function (app) {
...
 
-      for (var i = 0; i < len || i < oldLen; i++) {
-        patch(element, element.childNodes[i], oldNode.children[i], node.children[i])
-      }
+    var newIndex = 0
+    var oldIndex = 0
+    var cache = {}
+
+    while (
+      newIndex < newNode.children.length ||
```
Comment (jbucaran) [inline src/app.js]: @zaceno We should try to access stuff only once. `len`, `oldLen`, `oldNodeData`, etc.

File: src/app.js [hunk 15]
```diff
@@ -211,40 +213,107 @@ export default function (app) {
...
+        element.insertBefore(i[0], oldChildren[oldIndex][0])
+        patch(element, i[0], i[1], newChild)
+        delete cache[newChild.data.key]
+        newIndex++
+
+      } else if (oldChild == null || (isKeyed(oldChild) && !isKeyed(newChild))) {
+        patch(element, oldChildren[oldIndex][0], null, newChild)
+        newIndex++
+
+      } else if (newChild == null || (!isKeyed(oldChild) && isKeyed(newChild)) ) {
```
Comment (jbucaran) [inline src/app.js]: @zaceno Perhaps you can store the oldKey and newKey in a variable, then just check `oldKey != null` to see if a node is keyed or not.

File: src/app.js [hunk 16]
```diff
@@ -211,40 +213,107 @@ export default function (app) {
...
+      } else {
+        cache[oldChild.data.key] = oldChildren[oldIndex]
+        oldIndex++
+
+      }
     }
 
-    return element
+    for (var key in cache) {
+      batchRemove(element, cache[key][0], cache[key][1])
```
Comment (jbucaran) [inline src/app.js]: Noooice!

File: src/app.js [hunk 17]
```diff
@@ -211,40 +213,107 @@ export default function (app) {
...
+    return element
+  }
+
+  function patchChildren(element, oldNode, newNode) {
+    // We need to index all keyed nodes in oldChildren in order to
+    // distinguish new keyed nodes from nodes to reuse in newChildren
+    var oldChildren = []
+    var oldKeys = {}
+    for (var i = 0; i < oldNode.children.length; i++) {
+      oldNode.children[i].data &&
```
Comment (jbucaran) [inline src/app.js]: @zaceno Let's if instead of && here. It adds no value since uglifyjs will do and undo with this code anyway.

File: src/app.js [hunk 18]
```diff
@@ -211,40 +213,107 @@ export default function (app) {
...
+      oldNode.children[i].data.key &&
+      (oldKeys[oldNode.children[i].data.key] = true);
+      oldChildren[i] = [element.childNodes[i], oldNode.children[i]]
+    }
 
-      var len = node.children.length
-      var oldLen = oldNode.children.length
+    var isKeyed = function (node) {
+      return node.data && node.data.key && oldKeys[node.data.key]
```
Comment (jbucaran) [inline src/app.js]: @zaceno This means keys can't be numbers, specifically the number 0. I don't have an issue with that, but is that what we want?

Comment (leeoniya) [inline src/app.js]: 👎

File: src/app.js (hunk 15 above)
Comment (zaceno) [inline src/app.js]: I tried that, but that change actually *added* 11 bytes. Either way I'll leave the change in the upcoming commit, so that you can see the result and perhaps help me find a shorter way to write it.

File: src/app.js (hunk 18 above)
Comment (zaceno) [inline src/app.js]: Actually I think we can have numbers. The number 0 as well. Not sure exactly how various browsers handle this, but in v8 at least (just popped open my node console and tested this)

    > var a = {}
    undefined
    > a[0] = 'foo'
    'foo'
    > a 
    {'0': 'foo'}
    > a[0]
    'foo'

Sure, it will *cast* numbers to strings for the representation. But as long as you are not using a mix of numbers and stringed numbers to key your children (and *that* is a really unlikely scenario), it should work without conflict

File: src/app.js (hunk 15 above)
Comment (jbucaran) [inline src/app.js]: Interesting! Thanks for letting me know that :)

File: src/app.js (hunk 18 above)
Comment (jbucaran) [inline src/app.js]: @leeoniya Which one?

Comment (jbucaran) [inline src/app.js]: @zaceno If node.data.key is the number 0, then the above condition will result in iskeyed returning false.

What I meant was, if we are going to support keys as `Number`s, then the smallest check that works is using `null`.

Comment (leeoniya) [inline src/app.js]: right the condition will fail for all falsey values,which is wrong; all strings and numbers are valid keys. it's not a hash issue, which wull stringify all object keys.

File: src/app.js (hunk 15 above)
Comment (zaceno) [inline src/app.js]: I managed to compensate for those 11 bytes with some other changes (in the upcoming commit)

File: src/app.js (hunk 18 above)
Comment (zaceno) [inline src/app.js]: @leeoniya & @jbucaran: good points. Fixed in upcoming

**Comment by zaceno** [comment]
Current bundle size: 1935 bytes

**Comment by jbucaran** [comment]
@zaceno Now that #172 is out there, perhaps you can help there? We've been through this over Slack arleady. I am super thankful for all the work that went into this PR and all your feedback since you created #117 and I look forward more from you after this and from now on too! 👋

**Comment by zaceno** [comment]
@jbucaran Of course! I've already left a couple of comments and am continuing to review it. As far as code style goes, I haven't got a lot to say since it's your project :). But I am going to pull it down and run it through some tests later tonight.

Anyhow until 172 is merged, I'll try to keep 141 in parity. If only for the sake of personal pride (Since you mentioned it would be pulled into the history with 172 on top of it, I don't want my name attached to something half-assed ;) ) Removing the deferred is one such way. It was a very small change really -- took all of 5 minutes, so...

**Comment by zaceno** [comment]
Cleaned up the PR. Removed tests (as requested) and squashed to a single commit.

**Comment by jbucaran** [comment]
@zaceno Cool! I benchmarked my implementation and it's not as bad as I thought, 1.72, but I am sure we can do better.

Cache the old keys and old element lists might make a difference. Cache everything, we can even memoize toLowerCase. Also, going to try a different approach at removing stuff.

It would be great if you could review my PR and suggest more changes or how to use some of the features from your PR, since they are both pretty similar after all.

**Comment by zaceno** [comment]
@jbucaran Actually I suspect that our attempts to make it smaller have led to poorer performance. I've been hearing a lot of chatter lately about how specific declarative checks like `if (typeof foo === undefined)` run much faster than `if (!foo)` or `if (foo != null)` -- because of the casting that isn't necessary in the first case. Here's an example of such chatter I've been hearing 

https://github.com/developit/preact/pull/610

(I'll duplicate this post over on your PR so the discussion can continue there)

**Comment by jbucaran** [comment]
@zaceno 🎉
